### ISSUE TITLE
Won't run correctly when launched outside of a X terminal

### ISSUE THREAD

**Issue by Link-Satonaka** [cdown/clipmenu#33]
I'm certain this is an issue with my environment, but recently clipmenud stopped working with my systems user service. Specifically, it runs, but while it is running none of the selection buffers contain anything and I cannot copy or paste anything. `xsel -o` outputs nothing and xclip actually gives an error.

Running clipmenud from .xinitrc has the same issues. Running clipmenud from a terminal in X works fine.

I've gone through /proc/$pid/environ when clipmenud is working, and I tried to put all of those variables in the systemd unit. Didn't help.

I'm on archlinux- anyone have any idea what's happening?

**Comment by cdown** [comment]
When a non-working instance is running, what's the output of `tr '\0' '\n' < /proc/$$/environ | grep '^DISPLAY='`?

I suspect you need to do something like this in your .xinitrc:

    systemctl --user import-environment DISPLAY

**Comment by Link-Satonaka** [comment]
Yeah, the output of that command is DISPLAY=:0

My xinitrc sources `/etc/X11/xinit/xinitrc.d/50-systemd-user.sh`

```bash
#!/bin/sh

systemctl --user import-environment DISPLAY XAUTHORITY

if which dbus-update-activation-environment >/dev/null 2>&1; then
        dbus-update-activation-environment DISPLAY XAUTHORITY
fi
```

**Comment by Link-Satonaka** [comment]
I don't understand these results, but I did some debugging that might be useful to you.
Alterations to clipmenud
```diff
#!/bin/bash

cache_dir=/tmp/clipmenu.$USER/
mkdir -p -m0700 "$cache_dir"

declare -A last_data

while sleep "${CLIPMENUD_SLEEP:-0.5}"; do
    for selection in clipboard primary; do
+	echo "for selection" && sleep 5
        if type -p xsel >/dev/null 2>&1; then
            data=$(xsel --"$selection"; printf x)
        else
            data=$(xclip -o -sel "$selection"; printf x)
        fi
+	echo "post if" && sleep 5

        # We add and remove the x so that trailing newlines are not stripped.
        # Otherwise, they would be stripped by the very nature of how POSIX
        # defines command substitution.
        data=${data%x}

        [[ $data == *[^[:blank:]]* ]] || continue

        [[ ${last_data[$selection]} == "$data" ]] && continue
        last_data[$selection]=$data

        md5=$(md5sum <<< "$data")
        md5=${md5%% *}

        printf '%s' "$data" > "$cache_dir/$(LC_ALL=C date +%F-%H-%M-%S)-$md5"
    done
done
```
Loop to see script output and what's happening to the selection buffer.



```bash
while true; do systemctl --user status clipmenud.service --no-pager && echo "selection buffer:" `xsel -o` && sleep 2; done
```

Video of the results
https://ipfs.io/ipfs/QmQ7a8c4No59TShF3tGW4LzNbZTcdw2GH5yC7JXyG69WKL/output.mp4

**Comment by Link-Satonaka** [comment]
Basically, clipmenu works fine for a few iterations of the main while loop, then something erases the selection buffer.

**Comment by kaihendry** [comment]
I wasn't able to reproduce with `xclip` http://s.natalian.org/2016-10-31/1477879733_2548x1380.png

**Comment by Link-Satonaka** [comment]
Someone more knowledgeable than I should look through my dotfiles (zprofile/zshrc/xinitrc), because this issue is following me across my devices.

I've definitely stumbled across a bug in *something*, not necessarily clipmenu.

**Comment by cdown** [comment]
e8ebe8b adds a bunch of debug logging, can you see how this looks if you run with `DEBUG=1` in the environment?

This will print a lot, so you probably want to put output in a file.

**Comment by Link-Satonaka** [comment]
https://gist.github.com/Link-Satonaka/557611abea09acae1daa1a31492652d3

**Comment by Link-Satonaka** [comment]
My first time trying to wrangle journalctl / systemd logs, let me know if this came out all wrong.

Produced with this command
`journalctl --user-unit=clipmenud.service -S "today" -o cat > clip.log`

**Comment by Link-Satonaka** [comment]
Here's a simple log of clipmenu working fine, run from with my terminal.
```bash
DEBUG=1 clipmenud > clipmenworking.log 2>&1
```
https://gist.github.com/Link-Satonaka/5f155e7d77fb3eb9caa62ecaf0c6579c

**Comment by cdown** [comment]
```
% grep 'Data before stripping' /tmp/uu | sort | uniq -c | sort -r
    216 Data before stripping: x
      3 Data before stripping: Link@Talimx
      2 Data before stripping:
      1 Data before stripping: WantedBy=x11.targetx
      1 Data before stripping: DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/busx
      1 Data before stripping: Controls where file descriptor 1 (STDOUT) of the executed processes is connected to. Takes one of inherit, null, tty, journal, syslog, kmsg, journal+console, syslog+console, kmsg+console, socket or fd.x
% grep 'Skipping as' /tmp/uu | sort | uniq -c | sort -r 
    216 Skipping as clipboard is only blank
      1 Skipping as last selection is the same as this one
% grep 'possible partial' /tmp/uu | sort | uniq -c | sort -r 
% grep '^Writing' /tmp/uu | sort | uniq -c | sort -r 
      2 Writing
      1 Writing WantedBy=x11.target to /tmp/clipmenu.Link//2016-11-04-16:50:28.401107224
      1 Writing Link@Talim to /tmp/clipmenu.Link//2016-11-04-16:50:55.133807041
      1 Writing Link@Talim to /tmp/clipmenu.Link//2016-11-04-16:50:48.696041847
      1 Writing DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus to /tmp/clipmenu.Link//2016-11-04-16:50:13.954344556
      1 Writing Controls where file descriptor 1 (STDOUT) of the executed processes is connected to. Takes one of inherit, null, tty, journal, syslog, kmsg, journal+console, syslog+console, kmsg+console, socket or fd. to /tmp/clipmenu.Link//2016-11-04-16:49:24.203207024
```

Almost all of these are being skipped because apparently the clipboard is blank. At what point does it stop working?

**Comment by cdown** [comment]
Contrast that with your non-service:

```
% grep 'Data before stripping' /tmp/ur | sort | uniq -c | sort -r                                                                                                       
     23 Data before stripping: Administratorsx
     21 Data before stripping: complexx
      3 Data before stripping: examplex
      2 Data before stripping: seex
      2 Data before stripping: morex
      1 Data before stripping: frequentlyx
% grep 'Skipping as' /tmp/ur | sort | uniq -c | sort -r          
     44 Skipping as last selection is the same as this one
```

Something in your systemd environment keeps on wiping the clipboard

**Comment by Link-Satonaka** [comment]
Instantly. At no point is it working once I start the service.

In my alteration/debugging I provided earlier (where I inserted a few `sleep 3` statements), it only took a few iterations of the clipmenu script for the clipboard to clear. Here, without those pauses, the loop iterates fast enough where the clipboard is just instantly cleared before I can use it.

**Comment by cdown** [comment]
Well, as a systemd service it is working at least partially, clips appear but then almost immediately the clipboard is wiped by some unknown force :-)

I'll try to repro again today.

**Comment by cdown** [comment]
My suggestion would be to use an eBPF script + uprobes to analyse stuff calling X clipboard functions if I can't repro

**Comment by Link-Satonaka** [comment]
I added this line to my xinitrc; the issue still happens independent of systemd. I am way out of my understanding with what could possibly be happening.

`DEBUG=1 clipmenud > clipmenu_xinitrc.log 2>&1 &`

https://gist.github.com/Link-Satonaka/db2eae54e684c724d5510e3a303dbe0e

**Comment by cdown** [comment]
I'm really not sure what's going on here, but nothing major looks different in the environment between the working and unworking versions:

```
% diff -u /tmp/env-{terminal,service}
--- /tmp/env-terminal	2016-11-04 15:37:06.887427924 -0600
+++ /tmp/env-service	2016-11-04 15:36:29.750968345 -0600
@@ -1,48 +1,17 @@
-COLORTERM=yes
 DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
 DEBUG=1
 DISPLAY=:0
-DMENU_COL_CYAN=#005577
-DMENU_COL_GRAY1=#222222
-DMENU_COL_GRAY2=#444444
-DMENU_COL_GRAY3=#bbbbbb
-DMENU_COL_GRAY4=#eeeeee
-DMENU_FONT=inconsolata:size=8
-EDITOR=vim
-GTK_MODULES=canberra-gtk-module
 HOME=/home/Link
+JOURNAL_STREAM=8:88495
 LANG=en_US.UTF-8
-LESS=-R
-LESS_TERMCAP_mb=
-LESS_TERMCAP_md=
-LESS_TERMCAP_me=
-LESS_TERMCAP_se=
-LESS_TERMCAP_so=
-LESS_TERMCAP_ue=
-LESS_TERMCAP_us=
 LOGNAME=Link
-LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
 MAIL=/var/spool/mail/Link
-MMHEDITOR=/home/Link/.bin/vimeditmail
-MMHP=/home/Link/.mmh-profile
-MMHPAGER=/home/Link/.bin/vimpagermail
-MOZ_PLUGIN_PATH=/usr/lib/mozilla/plugins
-PAGER=less
-PATH=/home/Link/.bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
+MANAGERPID=486
+PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin
 PWD=/home/Link
-QT_STYLE_OVERRIDE=GTK+
 SHELL=/bin/zsh
-SHLVL=4
-SSH_AUTH_SOCK=/run/user/1000/ssh-agent.socket
-SUDO_ASKPASS=/home/Link/.bin/daskpass
-TERM=st-256color
-TERMCMD=st
+SHLVL=1
 USER=Link
-WINDOWID=20971526
-WINDOWPATH=1
 XAUTHORITY=/home/Link/.Xauthority
 XDG_RUNTIME_DIR=/run/user/1000
-XDG_SEAT=seat0
-XDG_SESSION_ID=c1
-XDG_VTNR=1
-_=/usr/bin/env
+_=/usr/sbin/env
% diff -u /tmp/env-{terminal,xinitrc}
--- /tmp/env-terminal	2016-11-04 15:37:06.887427924 -0600
+++ /tmp/env-xinitrc	2016-11-04 15:36:50.750850835 -0600
@@ -1,48 +1,30 @@
-COLORTERM=yes
 DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
 DEBUG=1
 DISPLAY=:0
-DMENU_COL_CYAN=#005577
-DMENU_COL_GRAY1=#222222
-DMENU_COL_GRAY2=#444444
-DMENU_COL_GRAY3=#bbbbbb
-DMENU_COL_GRAY4=#eeeeee
-DMENU_FONT=inconsolata:size=8
-EDITOR=vim
 GTK_MODULES=canberra-gtk-module
 HOME=/home/Link
 LANG=en_US.UTF-8
 LESS=-R
-LESS_TERMCAP_mb=
-LESS_TERMCAP_md=
-LESS_TERMCAP_me=
-LESS_TERMCAP_se=
-LESS_TERMCAP_so=
-LESS_TERMCAP_ue=
-LESS_TERMCAP_us=
 LOGNAME=Link
-LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
 MAIL=/var/spool/mail/Link
 MMHEDITOR=/home/Link/.bin/vimeditmail
 MMHP=/home/Link/.mmh-profile
 MMHPAGER=/home/Link/.bin/vimpagermail
 MOZ_PLUGIN_PATH=/usr/lib/mozilla/plugins
-PAGER=less
 PATH=/home/Link/.bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
 PWD=/home/Link
 QT_STYLE_OVERRIDE=GTK+
 SHELL=/bin/zsh
-SHLVL=4
+SHLVL=3
 SSH_AUTH_SOCK=/run/user/1000/ssh-agent.socket
 SUDO_ASKPASS=/home/Link/.bin/daskpass
-TERM=st-256color
+TERM=linux
 TERMCMD=st
 USER=Link
-WINDOWID=20971526
 WINDOWPATH=1
 XAUTHORITY=/home/Link/.Xauthority
 XDG_RUNTIME_DIR=/run/user/1000
 XDG_SEAT=seat0
-XDG_SESSION_ID=c1
+XDG_SESSION_ID=c4
 XDG_VTNR=1
 _=/usr/bin/env
```

The cgroups change, but that's to be expected. It's not like we fail to connect to to the X server in the failing case, it's like we succeed but something weird happens where the clipboard keeps getting wiped

**Comment by Link-Satonaka** [comment]
I tried killing basically every process, changing my WM, etc, in an attempt to find some process causing the interference. No luck.

I should note, if I didn't already, that the clipboard/selection both work fine while clipmenud is not running (and continue to work fine if clipmenud is run from a terminal). Only when clipmenud is run from xinitrc or a service does this issue manifest.

I found that, while using xterm, the selection is actively deselected. If I manage to run `xclip -o` before the selection gets deselected, the output will be, as it should, the contents of the selection. Using xterm, I can see exactly when the clipboard is cleared. (st, by contrast, doesn't deselect the text when the clipboard is cleared).

**Comment by tyjak** [comment]
I have the exact same issue both on a virtualbox env and on my macbook air install. The issue started on an update but I can't figure out what was updatedat that broke clipmenu. I'm using i3w and clipmenud was started just before i3w in my .xinitrc file. Now it's work only if I start clipmenud in the terminal.

**Comment by cdown** [comment]
Just out of curiosity, does this still happen on 1.0.0 (d439032)?

I'm wondering if this somehow has something to do with the clipboard ownership code.

**Comment by Link-Satonaka** [comment]
Yes, still happening using that commit.

**Comment by Link-Satonaka** [comment]
I went through and tried testing various older commits.
This is the earliest functioning commit
700a6b27dc6f2c87a027259d77a558268154502e

Here is the commit which introduces the problem
dd9a04184665cca972f0fbaf5b8e1c07dd0b079a

**Comment by Link-Satonaka** [comment]
@tyjak glad I'm not alone. Could you verify that using commit https://github.com/cdown/clipmenu/commit/700a6b27dc6f2c87a027259d77a558268154502e removes the issue for you?

**Comment by cdown** [comment]
That's interesting, I can only guess why that should cause problems now. Since this is a new problem, I guess it is some interaction between managing both clipboards and something else which is running as part of systemd (or changes in X, but that seems less likely).

Do you happen to run any clipboard sync utilities like autocutsel, or run a desktop environment which might do such clipboard synchronisation? My guess is that maybe they are conflicting somehow.

I'd be interested to know if adding a small sleep between the clipboard syncs fixes this for you. If it does, it's probably a read-after-write problem in X somehow.

**Comment by cdown** [comment]
917392f adds this logic, I'm also still trying to repro.

**Comment by cdown** [comment]
Hm, I do repro something weird in urxvt, finally. It seems that when clipmenud is running as a systemd unit, in urxvt the selection keeps on getting deselected. :/

**Comment by cdown** [comment]
(but the clips do still go into clipmenu, though, so I don't repro that part)

**Comment by cdown** [comment]
That part is some issue with getting PRIMARY, independent of clipmenud. Even with just this unit file, it keeps on getting deselected in urxvt:

```
[Service]
Type=simple
ExecStart=/usr/bin/xclip -sel primary
```

**Comment by cdown** [comment]
This looks related: https://sourceforge.net/p/parcellite/bugs/133/

**Comment by cdown** [comment]
This is the change they are talking about:

```diff
commit f709ee2a88eee006cb9da5817c1bbd7021c3a453
Author: RickyRockRat <gpib@rickyrockrat.net>
Date:   Mon Sep 29 18:16:58 2014 +0000

    First round at fix of 137/129
    
    git-svn-id: svn://svn.code.sf.net/p/parcellite/code/trunk@533 b25b1718-469e-46af-be73-919d414b4ebe

diff --git ChangeLog ChangeLog
index 9c95b4a..8f266b4 100644
--- ChangeLog
+++ ChangeLog
@@ -1,4 +1,4 @@
-V1.1.9 Saturday, July 5, 2014
+V1.1.9 Monday, Sept 29, 2014
  +Fixed bug 120, position_history broken. We limit x & y with screen limit -100.
  +Fixed bug 124, Enter does not select entry.
  +Fixed bug 125, Unsolicited error message about xdotool.
@@ -8,6 +8,7 @@ V1.1.9 Saturday, July 5, 2014
  +Fixed Bug 123, right-click edit broken.
  +Fixed unsolicited error message on startup (related to bug 125).
  +Cleaned up many warnings. Note to self: use env CFLAGS=-Wall ./configure --prefix=/usr/local
+ +Initial attempt at fixing bugs 129/137 - unable to copy files in file manager.
  
 V1.1.8 Thursday, July 3, 2014
  +Fixed bug 107, segfault when action key missing in parcelliterc.
diff --git src/main.c src/main.c
index 466aed0..8b5c093 100644
--- src/main.c
+++ src/main.c
@@ -172,6 +172,8 @@ gchar *process_new_item(GtkClipboard *clip,gchar *ntext, int *mod)
 	glong len,nlen;
 	gchar *rtn=NULL;
 	int i=0;
+	if(NULL != mod)
+ 		*mod=0;
 	if(NULL == ntext)
 		return NULL;
 	
@@ -196,8 +198,7 @@ gchar *process_new_item(GtkClipboard *clip,gchar *ntext, int *mod)
 	}
 	/**set the clipboard to the last entry - effectively deleting this entry */
 	goto done;
- if(NULL != mod)
- 	*mod=0;
+
  process:  /**now process the text.  */
 	/*printf("proc\n"); fflush(NULL); */
   len=strlen(ntext);/*g_utf8_strlen(ntext,-1); */
@@ -247,9 +248,9 @@ gchar *_update_clipboard (GtkClipboard *clip, gchar *n, gchar **old, int set, in
 	if(NULL != n)	{
 #ifdef DEBUG_UPDATE
 	 	if(clip==primary)
-			g_printf("%03x-set PRI to '%s'\n",mode,n);
+			g_printf("%03x-set%d PRI to '%s'\n",mode,set,n);
 		else
-			g_printf("%03x-set CLI to '%s'\n",mode,n);
+			g_printf("%03x-set%d CLI to '%s'\n",mode,set,n);
 #endif
 		if( set)
 			gtk_clipboard_set_text(clip, n, -1);
@@ -424,10 +425,10 @@ gchar *update_clipboard(GtkClipboard *clip,gchar *intext,  gint mode)
 			/**only check processed/changed. No need to update this clip, since the text is already there.  
 			   If we have identical lines except for whitespace, we also have to set the clipboard.
 			*/
-			if(0 == p_strcmp(processed,*existing) && 0 == mod) set=0;
+			if(0 == p_strcmp(processed,changed) && 0 == mod) set=0;
 			else {
 				set=1; 
-				DTRACE(g_fprintf(stderr,"set=%d. p='%s'\n",set,processed));
+				DTRACE(g_fprintf(stderr,"set=%d. c='%s' p='%s'\n",set,changed,processed));
 				/*set=1; */ /** Always set the text.*/
 			}	
 			last=_update_clipboard(clip,processed,existing,set,mode|H_MODE_CHANGED_MASK);
@@ -460,10 +461,16 @@ gchar *update_clipboard(GtkClipboard *clip,gchar *intext,  gint mode)
 	*/		
 	
 	if(H_MODE_NEW==mode){
+		gchar *unproc_clip;
+		if( NULL != intext)
+			unproc_clip=g_strdup(intext);
+		else unproc_clip=g_strdup("_NoClip_xyzpdq_");
 		if(NULL != (processed=process_new_item(clip,intext,&mod)) ){
-/*			g_printf("%sNEW '%s'\n",clipname,processed); */
-			if(0 == p_strcmp(processed,*existing))set=0;
+			DTRACE(g_printf("%sNEW '%s' was '%s'\n",clip==clipboard?"CLI":"PRI",processed,*existing)); 
+			/**clipboard already has selection, no need to update if we did not change it via processing  */
+			if(0 == p_strcmp(processed,unproc_clip))set=0;
 			else set=1;
+			g_free(unproc_clip);
 			last=_update_clipboard(clip,processed,existing,set,mode);
 			if(NULL != last)
 				append_item(last,current_on_top?HIST_DEL|HIST_CHECKDUP|HIST_KEEP_FLAGS:0,0,CLIP_TYPE_TEXT);
diff --git src/preferences.c src/preferences.c
index c4d3207..4812eda 100644
--- src/preferences.c
+++ src/preferences.c
@@ -662,7 +662,7 @@ void read_preferences(int mode)
 {
 	gchar *c,*rc_file = g_build_filename(g_get_user_config_dir(), PREFERENCES_FILE, NULL);
   gint32 z;
-  GError *err;
+  GError *err=NULL;
 	struct pref_item *p;
 	init_pref();
   /* Create key */
```

**Comment by cdown** [comment]
Well, this also happens with xsel, so no luck there.

**Comment by cdown** [comment]
Oh man, deja vu: https://github.com/cdown/clipmenu/blob/05c505dbfbc9b6b0d0e8dfa4c990878934386fcd/clipmenud#L58-L64

**Comment by cdown** [comment]
I worked at least this selection issue out -- I think this is related, so I'm hoping it also fixes your issue. Hooray!

`xsel` without `-o` even though it has the same *end* result, has different internal logic. Specifically, it seems to do things with the PRIMARY selection that cause weird behaviour.

I'm not sure why this only manifests when launched as a systemd unit, but this fixes the problem for me.

Try out c5f101d and let me know whether you still see problems :-)

**Comment by Link-Satonaka** [comment]
Success! Thank you so much!

Any idea what could have changed in the past month or so- why did this suddenly become an issue now?

**Comment by cdown** [comment]
Absolutely no idea. I need to look at what `-o` and non `-o` do differently, then we can go from there. Bizarre stuff :-)

**Comment by cdown** [comment]
This is included in 1.2.0, which was just released.
